test/test_memoize_decorator.py

```diff
@@ -927,3 +927,90 @@ def foo(bar=1) -> None:
         ...
 
     foo()
+
+
+def test_object_hash_produces_bad_hashes() -> None:
```

Comment:
This test demonstrates the problem, but probably isn't suitable to be included as part of our test suite. The non-unique hash affects my workstation (Python 3.7.4; Windows 10 and ubuntu 18.04) and my containerised builds (`FROM python:3.7-alpine`), but clearly doesn't affect our Travis builds (Python 3.7.1; Ubuntu 16.04)

Reply:
Ah, I see what the problem is now.

I wonder if we could change the way the default hash works. Currently, we create a tuple of keyword/arg pairs given to the function and call has on it ([code here](https://github.com/cevans87/atools/blob/master/atools/_memoize_decorator.py#L127)). If we were to not call hash and instead use the tuple itself, we'd still have a valid key for memoize. The problem with that is that by storing all the arguments in that tuple, we would keep the arguments from being garbage collected once all other references to those arguments were deleted and we'd get a memory leak. I actually did exactly this for a while before I realized I had created the memory leak (test [here](https://github.com/cevans87/atools/blob/master/test/test_memoize_decorator.py#L561)).

What if, instead we use the tuple again, but only store weakrefs to all the argument values? If we have a cache hit, we could actually iterate through those argument values and ensure that the weak references are still valid, and are exactly the same values as before. So, even if we ended up with with an argument that is an instance of an object that had a recycled id, we'd see that our memoized key has a dead weakref and we'd know that we ran into the case you pointed out here.

Let me know if that isn't very clear. I can write up a code snippet if needed.